File: item-list/src/main/java/com/gurunars/item_list/Differ.java
package com.gurunars.item_list;

import android.util.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import java8.util.function.BiFunction;


class Differ<ItemType extends Item> implements BiFunction<List<ItemType>, List<ItemType>, List<Change<ItemType>>> {

    @Inject
    private
    BiFunction<
        List<ItemType>,
        List<ItemType>,
        List<ItemType>> diffFetcher =
            new OrderedDiffFetcher<>();

    @Inject
    private
    FetcherPermutations<ItemType> fetcherPermutations = new FetcherPermutations<>();

    @Inject
    private Partitioner<ItemType> partitioner = new Partitioner<>();

    private void verifyNoDuplicates(List<ItemType> items) {
        Set<ItemType> set = new HashSet<>(items);
        if (set.size() != items.size()) {
            throw new RuntimeException("The list of items contains duplicates");
        }
    }

    private List<Change<ItemType>> getUpdates(
            List<ItemType> sourceList, List<ItemType> targetList, int offset) {
        List<Change<ItemType>> changes = new ArrayList<>();

        for (int i=0; i < sourceList.size(); i++) {
            ItemType newItem = targetList.get(i);
            int realIndex = offset + i;
            if (!sourceList.get(i).payloadsEqual(newItem)) {
                changes.add(new ChangeUpdate<>(newItem, realIndex, realIndex));
            }
        }

        return changes;
    }

    private List<ItemType> reverse(List<ItemType> original) {
        Collections.reverse(original);
        return original;
    }

    @Nonnull
    @Override
    public List<Change<ItemType>> apply(List<ItemType> source, List<ItemType> target) {

        verifyNoDuplicates(target);

        TmpTiming timing = new TmpTiming();

        Partitioner.PartitionTuple<ItemType> tuple = partitioner.apply(source, target);

        List<ItemType> sourceMiddle = tuple.getSource().getMiddle();
        List<ItemType> targetMiddle = tuple.getTarget().getMiddle();

        Log.e("SRC", "" + sourceMiddle);
        Log.e("TRG", "" + targetMiddle);

        List<Change<ItemType>> changes = new ArrayList<>();

        //timing.tick("INIT");

        // remove in a reverse order to prevent index recalculation
        for (ItemType item: reverse(diffFetcher.apply(sourceMiddle, targetMiddle))) {
            int position = sourceMiddle.indexOf(item);
            int realIndex = tuple.getStartOffset() + position;
            changes.add(new ChangeDelete<>(item, realIndex, realIndex));
            sourceMiddle.remove(position);
        }

        //timing.tick("REMOVALS");

        for (ItemType item: diffFetcher.apply(targetMiddle, sourceMiddle)) {
            int position = targetMiddle.indexOf(item);
            int realIndex = tuple.getStartOffset() + position;
            changes.add(new ChangeCreate<>(item, realIndex, realIndex));
            sourceMiddle.add(position, item);
        }

        //timing.tick("ADDITIONS");

        // Fetch permutations in both
        changes.addAll(fetcherPermutations.get(sourceMiddle, targetMiddle));

        //timing.tick("MOVES");

        changes.addAll(getUpdates(
                tuple.getSource().getHead(),
                tuple.getTarget().getHead(),
                0)
        );

        Log.e("SRC 2", "" + sourceMiddle);
        Log.e("TRG 2", "" + targetMiddle);

        changes.addAll(getUpdates(
                sourceMiddle,
                targetMiddle,
                tuple.getStartOffset())
        );
        changes.addAll(getUpdates(
                tuple.getSource().getTail(),
                tuple.getTarget().getTail(),
                tuple.getStartOffset() + sourceMiddle.size())
        );

        //timing.tick("UPDATES");

        return changes;
    }
}

File: item-list/src/main/java/com/gurunars/item_list/Differ.kt
package com.gurunars.item_list

import java.util.ArrayList
import java.util.Collections
import java.util.HashSet


internal class Differ<ItemType : Item> : BiFunction<List<ItemType>, List<ItemType>, List<Change<ItemType>>> {

    private val diffFetcher = OrderedDiffFetcher<ItemType>()
    private val fetcherPermutations = FetcherPermutations<ItemType>()
    private val partitioner = Partitioner<ItemType>()
    private val plainUpdateFetcher = PlainUpdateFetcher<ItemType>()
    private val mutatedUpdateFetcher = MutatedUpdateFetcher<ItemType>()

    private fun verifyNoDuplicates(items: List<ItemType>) {
        val set = HashSet(items)
        if (set.size != items.size) {
            throw RuntimeException("The list of items contains duplicates")
        }
    }

    private fun reverse(original: List<ItemType>): List<ItemType> {
        Collections.reverse(original)
        return original
    }

    override fun apply(source: List<ItemType>, target: List<ItemType>): List<Change<ItemType>> {

        verifyNoDuplicates(target)

        val tuple = partitioner.apply(source, target)

        val sourceMiddle = tuple.source.middle.toMutableList()
        val targetMiddle = tuple.target.middle

        val changes = ArrayList<Change<ItemType>>()

        // remove in a reverse order to prevent index recalculation
        for (item in reverse(diffFetcher.apply(sourceMiddle, targetMiddle))) {
            val position = sourceMiddle.indexOf(item)
            val realIndex = tuple.startOffset + position
            changes.add(ChangeDelete(item, realIndex, realIndex))
            sourceMiddle.removeAt(position)
        }

        for (item in diffFetcher.apply(targetMiddle, sourceMiddle)) {
            val position = targetMiddle.indexOf(item)
            val realIndex = tuple.startOffset + position
            changes.add(ChangeCreate(item, realIndex, realIndex))
            sourceMiddle.add(position, item)
        }

        // Fetch permutations in both
        changes.addAll(fetcherPermutations.get(sourceMiddle, targetMiddle, tuple.startOffset))
        changes.addAll(plainUpdateFetcher[tuple.source.head, tuple.target.head, 0])
        changes.addAll(mutatedUpdateFetcher.get(sourceMiddle, targetMiddle, tuple.startOffset))
        changes.addAll(plainUpdateFetcher[tuple.source.tail, tuple.target.tail,
                       tuple.startOffset + sourceMiddle.size])

        return changes
    }

}

